rust/src/shards/gui/menus/menu.rs

```diff
@@ -1,399 +1,413 @@
-    }
-
-    Ok(common_type::bool)
-  }
-
-  fn warmup(&mut self, ctx: &Context) -> Result<(), &str> {
-    self.parents.warmup(ctx);
-    if !self.contents.is_empty() {
-      self.contents.warmup(ctx)?;
-    }
-
-    Ok(())
-  }
-
-  fn cleanup(&mut self) -> Result<(), &str> {
-    if !self.contents.is_empty() {
-      self.contents.cleanup();
-    }
-    self.parents.cleanup();
-
-    Ok(())
-  }
-
-  fn activate(&mut self, context: &Context, input: &Var) -> Result<Var, &str> {
-    if self.contents.is_empty() {
-      // no contents, same as inactive
-      return Ok(false.into());
-    }
-
-    if let Some(ui) = util::get_current_parent(*self.parents.get())? {
-      match egui::menu::bar(ui, |ui| {
-        util::activate_ui_contents(context, input, ui, &mut self.parents, &mut self.contents)
-      })
-      .inner
-      {
-        // propagate the error
-        Err(err) => Err(err),
-        // menubar is active
-        Ok(_) => Ok(true.into()),
-      }
-    } else {
-      Err("No UI parent")
-    }
-  }
-}
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright © 2022 Fragcolor Pte. Ltd. */
+
+use super::CloseMenu;
+use super::Menu;
+use super::MenuBar;
+use crate::shard::Shard;
+use crate::shards::gui::util;
+use crate::shards::gui::HELP_OUTPUT_EQUAL_INPUT;
+use crate::shards::gui::HELP_VALUE_IGNORED;
+use crate::shards::gui::PARENTS_UI_NAME;
+use crate::types::common_type;
+use crate::types::Context;
+use crate::types::ExposedTypes;
+use crate::types::InstanceData;
+use crate::types::OptionalString;
+use crate::types::ParamVar;
+use crate::types::Parameters;
+use crate::types::ShardsVar;
+use crate::types::Type;
+use crate::types::Types;
+use crate::types::Var;
+use crate::types::ANY_TYPES;
+use crate::types::BOOL_TYPES;
+use crate::types::SHARDS_OR_NONE_TYPES;
+use crate::types::STRING_TYPES;
+
+lazy_static! {
+  static ref MENU_PARAMETERS: Parameters = vec![
+    (
+      cstr!("Title"),
+      shccstr!("The title of the menu."),
+      &STRING_TYPES[..],
+    )
+      .into(),
+    (
+      cstr!("Contents"),
+      shccstr!("The UI contents."),
+      &SHARDS_OR_NONE_TYPES[..],
+    )
+      .into(),
+  ];
+  static ref MENUBAR_PARAMETERS: Parameters = vec![(
+    cstr!("Contents"),
+    shccstr!("The UI contents."),
+    &SHARDS_OR_NONE_TYPES[..],
+  )
+    .into(),];
+}
+
+impl Default for CloseMenu {
+  fn default() -> Self {
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      parents,
+      requiring: Vec::new(),
+    }
+  }
+}
+
+impl Shard for CloseMenu {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.CloseMenu")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.CloseMenu-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.CloseMenu"
+  }
+
+  fn help(&mut self) -> OptionalString {
+    OptionalString(shccstr!("Closes the currently opened menu."))
```

Comment:
![Untitled](https://user-images.githubusercontent.com/109656055/208394250-0a2d7f92-e22a-43bf-b278-d39229995686.png)

The text is not showing for some reason.

Reply:
It's showing for me. Did you rebuild `shards` in debug before generating the doc?

Reply:
Ah yes that was the problem. I see it now, thanks!